docs/creating-sql-with-python.md

```diff
@@ -0,0 +1,201 @@
+to develop a faster methodology for adding new datasets.
+
+Occasionally you will have a dataset from a third party that is useful for working
+with your dataset. In the vocab study (requiring a license to use), we 
+[add coding system data](../cumulus_library/studies/vocab/vocab_icd_builder.py)
+from flat files to athena. If you need to do this, you should extend the base
+TableBuilder class, and your `prepare_queries` function should do the following,
+leveraging the
+[template function library](../cumulus_library/template_sql/templates.py):
+- Use the `get_ctas_query` function to get a CREATE TABLE AS statement to 
+instantiate your table in athena
+- Since athena SQL queries are limited in size to 262144 bytes, if you have
+a large dataset, break it up into smaller chunks
+- Use the `get_insert_into` function to add the data from each table to
+the chunk you just created.
+
+Add the dataset uploader to the `table_builder_config` section of your
+`manifest.toml` to include it in your build - this will make this data
+available for downstream queries
+
+### Querying nested data
+
+Are you trying to access data from deep within raw FHIR tables? I'm so sorry.
+Here's an example of how this can get fussy with code systems:
+
+A FHIR coding element may be an array, or it may be a singleton, or it may
+be a singleton wrapped an array. It may be fully populated, or partially populated,
+or completely absent. There may be one code per record, or multiple codes per record,
+and you may only be interested in a subset of these codes.
+
+This means you may have differing schemas in Athena from one site's data to another
+(especially if they come from different EHR systems, where implementation details
+may differ). In order to handle this, you need to create a standard output
+representation that accounts for all the different permutations you have, and
+conform data to match that. The 
+[encounter coding](../cumulus_library/studies/core/builder_encounter_coding.py)
+and
+[condition codeableConcept](../cumulus_library/studies/core/builder_condition_codeableconcept.py)
+builders both jump through hoops to try and get this data into flat tables for
+downstream use.
+
+This is a pretty open ended design problem, but based on our experience, your
+`prepare_queries` implementation should attempt the following steps:
+- Check if your table has any data at all
+- If it does, inspect the table schema to see if the field you're interested in
+is populated with the schema elements you're expecting
+  - If yes, it's safe to grab them
+  - If no, you will need to manually initialize them to an appropriate null value
+- If you are dealing with deeply nested objects, you may need to repeat the above
+more than once
+- Write a jinja template that handles the conditionally present data, and a 
+template function to invoke that template
+- Test this against as many different sample databases as you can
```

Comment:
nit: maybe "test against data samples from many different EHRs" or similar? "sample databases" feels like jargon from our team to mean different EHR samples, but the reader may not have that brain context.

Reply:
yeah i buy it